libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,336 @@
 
-        client.publishMessage("foo.bar", "bar.foo", "hello there");
+            assertNotNull(subscription1.nextMessage(TIMEOUT_MILLIS));
+            assertNotNull(subscription2.nextMessage(TIMEOUT_MILLIS));
 
-        Message message = fooSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello there", new String(message.getData()));
+            subscription1.unsubscribe();
+            subscription2.unsubscribe();
+        }
     }
 
     @Test
-    public void whenWildCardSubscription_andNotMatchTopic_NoMessageReceived() throws Exception {
-
-        NatsClient client = connectClient();
+    public void whenSubscribingSynchronouslyWithQueueGroups_thenOnlyOneSubscriberInTheGroupShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            Subscription qSub1 = natsConnection.subscribe("mySubject", "myQueue");
+            Subscription qSub2 = natsConnection.subscribe("mySubject", "myQueue");
 
-        SyncSubscription fooSubscription = client.subscribeSync("foo.*");
+            natsConnection.publish("mySubject", "data".getBytes());
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
+            List<Message> messages = new ArrayList<>();
 
-        Message message = fooSubscription.nextMessage(200);
-        assertNull("Got message!", message);
+            Message message = qSub1.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
 
+            message = qSub2.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
 
-        SyncSubscription barSubscription = client.subscribeSync("foo.>");
+            assertEquals(1, messages.size());
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
-
-        message = barSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello there", new String(message.getData()));
+            qSub1.unsubscribe();
+            qSub2.unsubscribe();
+        }
+    }
 
+    @Test
+    public void whenSubscribingAsynchronouslyWithoutQueueGroups_thenEachMessageHandlerShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            List<Message> messages1 = new ArrayList<>();
+            List<Message> messages2 = new ArrayList<>();
+
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            Subscription subscription1 = dispatcher.subscribe("mySubject", messages1::add);
+            Subscription subscription2 = dispatcher.subscribe("mySubject", messages2::add);
+
+            natsConnection.publish("mySubject", "data1".getBytes());
+            natsConnection.publish("mySubject", "data2".getBytes());
+
+            // Simulate real world time for messages to propagate.
```

Comment:
We aren't really simulating anything here, we are literally waiting for the message to arrive
```suggestion
            // wait for messages to propagate.
```

Reply:
fixed